### PR TITLE
module: fix re-entrant module importing

### PR DESCRIPTION
**Pull request by obastemur**

### CODE REVIEW

**Comment by obastemur** [comment]
Fixes #4570 . Test cases will follow. /cc @rhuanjl @fatcerberus

File: lib/Runtime/Language/SourceTextModuleRecord.cpp [hunk 1]
```diff
@@ -722,8 +722,8 @@ namespace Js
                     Recycler* recycler = GetScriptContext()->GetRecycler();
                     this->parentModuleList = RecyclerNew(recycler, ModuleRecordList, recycler);
                 }
-                bool contains = this->parentModuleList->Contains(parentRecord);
-                if (!contains)
+
+                if (!this->parentModuleList->Contains(parentRecord))
```
[the remark this answers is not in the document]
Comment (akroshg) [reply lib/Runtime/Language/SourceTextModuleRecord.cpp]: The assert was wrong - I have removed that.

---
In reply to: [162434163](https://github.com/Microsoft/ChakraCore/pull/4574#discussion_r162434163) [](ancestors = 162434163)

**Comment by obastemur** [comment]
Prior to adding more test cases, I will introduce `WScript.RegisterModuleSource(path, source)`. So we won't end-up having bunch of moduleXXX1.js files under test folder.

**Comment by akroshg** [comment]
that would be great!

---
In reply to: [358745237](https://github.com/Microsoft/ChakraCore/pull/4574#issuecomment-358745237) [](ancestors = 358745237)

**Comment by rhuanjl** [comment]
The test case I had is here:
https://github.com/rhuanjl/ChakraCore/commit/6b050fda8d42d56d3ca526d027e820feb7488333

Though I note that this issue would be covered by currently disabled test bug_issue_3257.js

**Comment by obastemur** [comment]
> Though I note that this issue would be covered by currently disabled test bug_issue_3257.js

It would do that to some extend. I saw there are other issues need attention prior to enable it. See https://github.com/Microsoft/ChakraCore/issues/4575

**Comment by obastemur** [comment]
Also opened https://github.com/Microsoft/ChakraCore/issues/4577 to track `WScript.RegisterModuleSource(path, source)` separately.

**Comment by obastemur** [comment]
@akroshg @boingoing please review

**Comment by rhuanjl** [comment]
This seems to work for most cases now.

The following seemingly related example (i.e. a circular re-entrant import) still crashes but I think it crashes with or without this fix - a few printfs show that this generating 5 module records - taking module0 out of the folder removes the crash so this is probably an error in ch's specifier normalising or something and not a CC error: 

```js
//test.js
WScript.LoadScriptFile("module0/module0.js","module");
```

```js
//module0/module0.js
import "module1.js";
import "module2.js";
```

```js
//module1.js
import "module2.js";
```

```js
//module2.js
import "module0/module0.js";
```

**Comment by obastemur** [comment]
Yep there are more to come. Things to do on both ch and CC side. Thanks for the test cases and all the effort @rhuanjl !

**Comment by obastemur** [comment]
@akroshg @rhuanjl @boingoing Thanks for the reviews!

**Comment by fatcerberus** [comment]
@obastemur This fix has resurfaced #4482 in miniSphere. :(

**Pull request merged**